lib/pure/socketstreams.nim

```diff
@@ -0,0 +1,171 @@
+## This module provides an implementation of the streams interface for sockets.
+## It contains two separate implementations, a
+## `ReadSocketStream <#ReadSocketStream>` and a
+## `WriteSocketStream <#WriteSocketStream>`. The `ReadSocketStream` only
+## supports reading, peeking, and seeking. It reads into a buffer, so even by
+## seeking backwards it will only read the same position a single time from the
+## underlying socket. To clear the buffer and free the data read into it you
+## can call `resetStream`, this will also reset the position back to 0 but
+## won't do anything to the underlying socket.
+##
+## The `WriteSocketStream` allows both reading and writing, but it performs the
+## reads on the internal buffer. So by writing to the buffer you can then read
+## back what was written but without receiving anything from the socket. You
+## can also set the position and overwrite parts of the buffer, and to send
+## anything over the socket you need to call `flush` at which point you can't
+## write anything to the buffer before the point of the flush (but it can still
+## be read). Again to empty the underlying buffer you need to call
+## `resetStream`.
+##
+## Examples
+## =======
+##
+## .. code-block:: Nim
+##  import socketstreams
+##
+##  var
+##    socket = newSocket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
+##    stream = newReadSocketStream(socket)
+##  socket.sendTo("127.0.0.1", Port(12345), "SOME REQUEST")
+##  echo stream.readLine() # Will call `recv`
+##  stream.setPosition(0)
+##  echo stream.readLine() # Will return the read line from the buffer
+##  stream.resetStream() # Buffer is now empty, position is 0
+##  echo stream.readLine() # Will call `recv` again
+##  stream.close() # Closes the socket
+##
+## .. code-block:: Nim
+##
+##  import socketstreams
+##
+##  var socket = newSocket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
+##  socket.connect("127.0.0.1", Port(12345))
+##  var sendStream = newWriteSocketStream(socket)
+##  sendStream.write "NOM"
+##  sendStream.setPosition(1)
+##  echo sendStream.peekStr(2) # OM
+##  sendStream.write "I"
+##  sendStream.setPosition(0)
+##  echo sendStream.readStr(3) # NIM
+##  echo sendStream.getPosition() # 3
+##  sendStream.flush() # This actually performs the writing to the socket
+##  sendStream.setPosition(1)
+##  sendStream.write "I" # Throws an error as we can't write into an already sent buffer
+
+import net, streams
+
+type
+  ReadSocketStream* = ref ReadSocketStreamObj
+  ReadSocketStreamObj* = object of StreamObj
+    data: Socket
+    pos: int
+    buf: seq[byte]
+  WriteSocketStream* = ref WriteSocketStreamObj
+  WriteSocketStreamObj* = object of ReadSocketStreamObj
+    lastFlush: int
+
+proc rsAtEnd(s: Stream): bool =
+  return false
+
+proc rsSetPosition(s: Stream, pos: int) =
+  var s = ReadSocketStream(s)
+  s.pos = pos
+
+proc rsGetPosition(s: Stream): int =
+  var s = ReadSocketStream(s)
+  return s.pos
+
+proc rsPeekData(s: Stream, buffer: pointer, bufLen: int): int =
```

Comment:
change with this:
```nim
proc rsPeekData(s: Stream, buffer: pointer, bufLen: int): int =
  let s = ReadSocketStream(s)
  if bufLen > 0:
    let oldLen = s.buf.len
    s.buf.setLen(max(s.pos + bufLen, s.buf.len))
    if s.pos > oldLen:
      result = s.data.recv(s.buf[s.pos].addr, s.pos - s.buf.len + bufLen)
    elif s.pos == oldLen:
      result = s.data.recv(s.buf[s.pos].addr, bufLen)
    elif s.pos + bufLen > oldLen:
      result = s.data.recv(s.buf[oldLen].addr, s.buf.len - oldLen)
    s.atEnd = result == 0
    copyMem(buffer, s.buf[s.pos].addr, bufLen)
```

Reply:
Hey, might still be wrong, its hard to reason with so many levels of abstraction

Reply:
> `s.atEnd = result == 0`

gives "'atEnd(s)' cannot be assigned to", i.e. "attempting to call undeclared routine: 'atEnd='"

Reply:
Edit: should be good now, but needs a test

Reply:
Fixed!

Now it works for your example, as well as for the examples where you manually set position (as mentioned on IRC).